feat: mark opportunities as done

amhislop opened the pull request:
- [x] I've checked that this isn't a new swag opportunity proposal.
- [x] I've checked that this isn't a duplicate pull request.

Resolves #151 

Enhancements:
- Adds a 'mark as complete' option to each opportunity. 
- Completed opportunities persist to local storage.
- Adds a 'Hide completed opportunities' checkbox to filter options

vikaspotluri123 commented on src/js/index.js:
```diff
@@ -20,16 +20,39 @@ const filterInput = document.querySelector('#filter');
 
+function setCompleteAttr(el) {
+	Array.from(document.querySelectorAll('.item'))
+		.find(item => item.contains(el)).dataset.completed = el.checked;
```
I think using a classname would be better here - `.classList.toggle('checked-off', el.checked)`

vikaspotluri123 commented on src/js/index.js:
```diff
@@ -64,14 +87,19 @@ function handleTags() {
 			tags.reduce((sho, tag) => sho || el.classList.contains(`tag-${tag}`), tags.length === 0);
-		if (!show) {
+
+		const hide = hideCompleted.checked && el.querySelector('.complete-notice').checked;
```
I think this logic can be tied together. However, I think we should definitely comment that different cases that affect visibility

vikaspotluri123 commented on src/styl/imports/card.styl:
```diff
@@ -69,3 +69,83 @@
+		cursor pointer
+
+		&::before
+			content 'Mark as complete?'
```
This isn't accessible. Also, this will compete for space with the `Date Added` coming in #328. I think this should be centered in a new row. I have a feeling #407 will move this metadata to below the picture, but don't worry too much about that

vikaspotluri123 commented:
Hey @amhislop 👋 thanks for contributing! I've left a couple of comments for now. Great job so far 🎉

amhislop commented:
Thanks for the feedback @vikaspotluri123 :thumbsup: I’ve made the below adjustments in the latest commit based on your suggestions.

- Switched to using a class name instead of data attribute for completed opportunities
- Merged the _show/hide_ conditions into single _show_ conditional and added comments
- Fixed the accessibility issue with the input checkbox field.  Centred the container on a new line as recommended. Checked layout of elements with `Date Added` element from #328 and the design looks nice and clean.

vikaspotluri123 commented:
Good work so far! I have more suggestions but I can't get to them now

amhislop commented:
Merged in the latest changes from the master branch and have tested them but am getting a valid image / reference notice in Travis CI. I'm not sure what I can do about this though.

aslafy-z commented:
Hello @amhislop! Any chance you can rebase this work on the new design (see #443)?
Thank you!

amhislop commented:
Hi @aslafy-z. Absolutely. No problem.

amhislop commented:
Hello @aslafy-z. I've rebased the pull request with the new design.

aslafy-z commented:
In fact it's not rebased on this repo's master branch. I guess you'll have to update your master branch first, and then rebase from it. Thanks for your time!

amhislop commented:
Sorry about that. I've fixed that up now.

aslafy-z commented on src/js/index.js:
```diff
@@ -62,8 +89,12 @@ function handleTags() {
-			tags.reduce((sho, tag) => sho || element.classList.contains(`tag-${tag}`), tags.length === 0);
+		const show = ((showExpired.checked || !element.classList.contains('tag-expired')) &&
+			tags.reduce((sho, tag) => sho || element.classList.contains(`tag-${tag}`), tags.length === 0)) && !(hideCompleted.checked && element.querySelector('.complete-notice').checked);
+
```
Please split this line in two like the one above.

aslafy-z commented:
Here's two remarks:
- The Done button eats a lot a space, I think we can find it a better place, and maybe move to label to a tooltip or something. @vikaspotluri123 @swapagarwal wdyt?
- When the hide complemented opportunities is enabled, if I set some opportunities as done, they stay in the list. They should disappear with some animation or something after few seconds (see the user can revert a missclick)

vikaspotluri123 commented:
I agree w/ both. Maybe have a bar / button at the top to toggle the state?

amhislop commented:
Thanks for the feedback, @aslafy-z and @vikaspotluri123. I've fixed the problem regarding marking opportunities as complete when the completed filter is checked and I've also added a fade out animation.

I've relocated the button to the top of the card and implemented a tooltip for each card, as suggested, using tippy.js.

aslafy-z commented on src/styl/imports/card.styl:
```diff
@@ -115,3 +120,89 @@
+		margin 1em
+		top 0
+		right 0
+
```
```suggestion
		z-index 999

```

aslafy-z commented:
Thank you for your awesome contribution @amhislop!
